src/script/socket.c

```diff
@@ -0,0 +1,224 @@
+	if (SOCKET_FAILED(socket)) {
+		ssock->socket = INVALID_SOCKET;
+		ssock->error = SocketError();
+	} else {
+		ssock->socket = socket;
+		ssock->error = 0;
+	}
+}
+
+struct mScriptValue* _mScriptSocketCreate() {
+	struct mScriptSocket client = {
+		.socket = INVALID_SOCKET,
+		.error = 0,
+		.port = 0
+	};
+
+	struct mScriptValue* result = mScriptValueAlloc(mSCRIPT_TYPE_MS_S(mScriptSocket));
+	result->value.opaque = calloc(1, sizeof(struct mScriptSocket));
+	*(struct mScriptSocket*) result->value.opaque = client;
+	return result;
+}
+
+void _mScriptSocketClose(struct mScriptSocket* ssock) {
+	if (!SOCKET_FAILED(ssock->socket)) {
+		SocketClose(ssock->socket);
+	}
+}
+
+struct mScriptValue* _mScriptSocketAccept(struct mScriptSocket* ssock) {
+	struct mScriptValue* value = _mScriptSocketCreate();
+	struct mScriptSocket* client = (struct mScriptSocket*) value->value.opaque;
+	_mScriptSocketSetSocket(client, SocketAccept(ssock->socket, &client->address));
+	if (!client->error) {
+		SocketSetBlocking(client->socket, false);
+	}
+	return value;
+}
+
+int32_t _mScriptSocketOpen(struct mScriptSocket* ssock, struct mScriptString* addressStr, uint16_t port) {
+	struct Address* addr = NULL;
+	if (addressStr && addressStr->length > 0) {
+		int32_t err = SocketResolveHost(addressStr->buffer, &ssock->address);
+		if (err) {
+			ssock->error = err;
+			return err;
+		}
+		addr = &ssock->address;
+	}
+	ssock->port = port;
+	_mScriptSocketSetSocket(ssock, SocketOpenTCP(port, addr));
+	if (!ssock->error) {
+		SocketSetBlocking(ssock->socket, false);
+	}
+	return ssock->error;
+}
+
+int32_t _mScriptSocketConnect(struct mScriptSocket* ssock, struct mScriptString* addressStr, uint16_t port) {
+	int32_t err = SocketResolveHost(addressStr->buffer, &ssock->address);
+	if (err) {
+		ssock->error = err;
+		return err;
+	}
+	ssock->port = port;
+	_mScriptSocketSetSocket(ssock, SocketConnectTCP(port, &ssock->address));
+	if (!ssock->error) {
+		SocketSetBlocking(ssock->socket, false);
+	}
+	return ssock->error;
+}
+
+int32_t _mScriptSocketListen(struct mScriptSocket* ssock, uint32_t queueLength) {
+	ssock->error = SocketListen(ssock->socket, queueLength);
+	return ssock->error;
+}
+
+int32_t _mScriptSocketSend(struct mScriptSocket* ssock, struct mScriptString* data) {
+	ssize_t written = SocketSend(ssock->socket, data->buffer, data->size);
+	if (written < 0) {
+		ssock->error = SocketError();
+		return -ssock->error;
+	}
+	ssock->error = 0;
+	return written;
+}
+
+struct mScriptValue* _mScriptSocketRecv(struct mScriptSocket* ssock, uint32_t maxBytes) {
+	struct mScriptValue* value = mScriptStringCreateEmpty(maxBytes);
+	struct mScriptString* data = value->value.string;
+	ssize_t bytes = SocketRecv(ssock->socket, data->buffer, maxBytes);
+	if (bytes <= 0) {
+		data->size = 0;
+		ssock->error = SocketError();
+	} else {
+		data->size = bytes;
+		ssock->error = 0;
+	}
+	return value;
+}
+
+// XXX: This is for prototyping. It's not a good idea.
```

Comment:
Is this comment still accurate?

Reply:
Largely, yes. It would be better to collect all of the active sockets and select on them in one go instead of having one select() call per active socket. It's not the worst thing in the world because I wouldn't expect one of these scripts to have a lot of sockets open, of course.

Reply:
Updated it.